File: data_structures/singly_linked_list.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(val) {
    // Add value at the end of the SLL
    var newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
      this.tail = this.head;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }

    this.length += 1;
    return this;
  }

  pop() {
    // Remove value from the end of the SLL
    if (!this.head) return undefined;
    var current = this.head;
    var newTail = current;
    while (current.next) {
      newTail = current;
      current = current.next;
    }
    this.tail = newTail;
    this.tail.next = null;
    this.length--;

    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }

    return current;
  }

  shift() {
    // Remove value from the start of the SLL
    if (!this.head) return undefined;
    var currentHead = this.head;
    this.head = currentHead.next;
    this.length--;
    if (this.length === 0) {
      this.tail = null;
    }
    return currentHead;
  }

  unshift(val) {
    // Add value at the start of the SLL
    var node = new Node(val);
    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head = node;
    }
    this.length++;
    return node;
  }

  get(index) {
    // Get node at the index of the SLL
    if (index < 0 || index > this.length || this.length === 0) return null;

    if (index === 0) {
      return this.head;
    } else if (index === this.length - 1) {
      return this.tail;
    }

    var currentPosition = 0;
    var currentNode = this.head;

    while (currentPosition <= index) {
      currentPosition++;
      currentNode = currentNode.next;
      if (currentPosition >= index) {
        return currentNode;
      }
    }
  }

  set(index, value) {
    // Set node at the index of the SLL
    var node = this.get(index);
    if (!node) return false;
    node.val = value;
    return true;
  }

  insert(index, val) {
    // Insert node at the index of the SLL
    if (index < 0 || index > this.length) return false;
    if (index === this.length) return !!this.push(val);
    if (index === 0) return !!this.unshift(val);

    var newNode = new Node(val);
    var prev = this.get(index - 1);
    var temp = prev.next;
    prev.next = newNode;
    newNode.next = temp;
    this.length++;

    return true;
  }

  remove(index) {
    if (index < 0 || index >= this.length) return undefined;
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();
    var previousNode = this.get(index - 1);
    var removed = previousNode.next;
    previousNode.next = removed.next;
    this.length--;
    return removed;
  }

  reverse() {
    var node = this.head;
    this.head = this.tail;
    this.tail = node;
    var next;
    var prev = null;
    for (var i = 0; i < this.length; i++) {
      next = node.next;
      node.next = prev;
      prev = node;
      node = next;
    }
    return this;
  }

  print() {
    var arr = [];
    var current = this.head
    while (current) {
      arr.push(current.val)
      current = current.next
    }
    console.log(arr);
  }
}

var list = new SinglyLinkedList();
list.push("Hello")
list.push("World")
list.push("!")

list.pop()
list.pop()

list.unshift("foo")
list.unshift("head!")

list.set(0, "set")

list.insert(1, "inserted")
list.insert(1, "inserted again")

list.remove(1)

var currentNode = list.head;

while (currentNode.next) {
  console.log(currentNode.val);
  if (!currentNode.next) {
    console.log(currentNode.next);
  }
  currentNode = currentNode.next;
}
console.log(list.tail.val);

File: data_structures/singly_linked_list.go
package main

import (
	"errors"
	"fmt"
)

// Node struct
type Node struct {
	val  int
	next *Node
}

// SinglyLinkedList struct
type SinglyLinkedList struct {
	length uint8
	head   *Node
	tail   *Node
}

// NewSinglyLinkedList is a SinglyLinkedList factory
func NewSinglyLinkedList() SinglyLinkedList {
	linkedList := SinglyLinkedList{}
	return linkedList
}

// SinglyLinkedList push add a new Node to the very end of a SLL
func (l *SinglyLinkedList) push(val int) *SinglyLinkedList {
	newNode := &Node{val: val}

	if l.head == nil {
		l.head = newNode
		l.tail = l.head
	} else {
		l.tail.next = newNode
		l.tail = newNode
	}

	l.length++
	return l
}

// SinglyLinkedList pop remove the last Node of a SLL
func (l *SinglyLinkedList) pop() (*Node, error) {
	if l.head == nil {
		return nil, errors.New("pop error")
	}

	current := l.head
	newTail := current

	for current.next != nil {
		newTail = current
		current = current.next
	}

	l.tail = newTail
	l.tail.next = nil
	l.length--

	if l.length == 0 {
		l.head = nil
		l.tail = nil
	}

	return current, nil
}

// SinglyLinkedList shift remove first Node of a SLL
func (l *SinglyLinkedList) shift() (*Node, error) {
	if l.head == nil {
		return nil, errors.New("shift error")
	}

	currentHead := l.head
	l.head = currentHead.next
	l.length--

	if l.length == 0 {
		l.tail = nil
	}

	return currentHead, nil
}

// SinglyLinkedList unshift remove first Node of a SLL
func (l *SinglyLinkedList) unshift(val int) *Node {
	node := &Node{val: val}

	if l.head == nil {
		l.head = node
		l.tail = node
	} else {
		node.next = l.head
		l.head = node
	}

	l.length++

	return node
}

// SinglyLinkedList unshift get Node at index of a SLL
func (l *SinglyLinkedList) get(index uint8) (*Node, error) {
	if index > l.length || l.length == 0 {
		return nil, errors.New("get could not be performed")
	}

	if index == 0 {
		return l.head, nil
	} else if index == l.length-1 {
		return l.tail, nil
	}

	var currentPosition uint8 = 0
	currentNode := l.head

	for currentPosition <= index {
		currentPosition++
		currentNode = currentNode.next
		if currentPosition >= index {
			break
		}
	}

	return currentNode, nil
}

// SinglyLinkedList set value at index of a SLL
func (l *SinglyLinkedList) set(index uint8, val int) bool {
	node, err := l.get(index)

	if err != nil {
		return false
	}

	node.val = val
	return true
}

// SinglyLinkedList insert a new Node between nodes of a SLL
func (l *SinglyLinkedList) insert(index uint8, val int) bool {
	if index > l.length {
		return false
	} else if index == l.length {
		l.push(val)
		return true
	} else if index == 0 {
		l.unshift(val)
		return true
	}

	node := &Node{val: val}
	prev, _ := l.get(index - 1)
	temp := prev.next
	prev.next = node
	node.next = temp
	l.length++

	return true
}

// SinglyLinkedList remove Node at index of a SLL
func (l *SinglyLinkedList) remove(index uint8) (*Node, error) {
	if index > l.length-1 {
		return nil, errors.New("error performing remove at index")
	} else if index == 0 {
		l.shift()
	} else if index == l.length-1 {
		l.pop()
	}

	prev, err := l.get(index - 1)

	if err != nil {
		return nil, err
	}

	node := prev.next
	prev.next = node.next
	l.length--

	return node, nil
}

// SinglyLinkedList reverse inverts a SLL
func (l *SinglyLinkedList) reverse() *SinglyLinkedList {
	node := l.head
	l.head = l.tail
	l.tail = node
	var next *Node
	var prev *Node

	for i := 1; uint8(i) < l.length; i++ {
		next = node.next
		node.next = prev
		prev = node
		node = next
	}

	return l
}

func log(statement string, l SinglyLinkedList) {
	fmt.Println(statement, "\n->", "Head:", l.head, "Tail:", l.tail, "Length:", l.length)
}

func main() {
	linkedList := NewSinglyLinkedList()

	linkedList.push(1)
	linkedList.push(2)
	linkedList.push(3)
	log("after_push 3x", linkedList)

	linkedList.pop()
	linkedList.pop()
	linkedList.pop()
	log("after_pop 3x", linkedList)

	linkedList.push(1)
	linkedList.push(2)
	linkedList.push(3)
	log("after_push 3x", linkedList)

	linkedList.shift()
	log("after_shift 1x", linkedList)

	linkedList.unshift(100)
	log("after_unshift 1x", linkedList)

	fmt.Println(linkedList.get(0))
	fmt.Println(linkedList.get(1))
	fmt.Println(linkedList.get(2))

	linkedList.set(0, 1)
	log("after_set at 0", linkedList)

	linkedList.set(0, 100)
	log("after_set at 0", linkedList)
	linkedList.set(1, 200)
	log("after_set at 1", linkedList)
	linkedList.set(2, 300)
	log("after_set at 2", linkedList)

	linkedList.insert(1, 150)
	log("after_insert at 1", linkedList)

	linkedList.remove(1)
	log("after_remove at 1", linkedList)

	fmt.Println(linkedList.get(1))

	linkedList.reverse()
	log("after_reverse", linkedList)
}
